cpp/unreal_plugins/SimulationController/Source/SimulationController/UrdfBotAgent.cpp

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+}
+
+UrdfBotAgent::~UrdfBotAgent()
+{
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "camera")) {
+        ASSERT(camera_sensor_);
+        camera_sensor_ = nullptr;
+    }
+
+    ASSERT(urdf_bot_pawn_);
+    urdf_bot_pawn_->Destroy();
+    urdf_bot_pawn_ = nullptr;
+}
+
+void UrdfBotAgent::findObjectReferences(UWorld* world) {}
+
+void UrdfBotAgent::cleanUpObjectReferences() {}
+
+std::map<std::string, Box> UrdfBotAgent::getActionSpace() const
+{
+    std::map<std::string, Box> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "joint")) {
+        for (auto& joint_component : urdf_bot_pawn_->urdf_robot_component_->joint_components_) {
+            if (joint_component.second->control_type_ != UrdfJointControlType::Invalid) {
+                Box box;
+                box.low_ = std::numeric_limits<float>::lowest();
+                box.high_ = std::numeric_limits<float>::max();
+                box.shape_ = {1};
+                box.datatype_ = DataType::Float32;
+                action_space[joint_component.first] = std::move(box);
+            }
+        }
+    }
+
+    return action_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getObservationSpace() const
+{
+    std::map<std::string, Box> observation_space;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "link_state")) {
+        for (auto& link_component : urdf_bot_pawn_->urdf_robot_component_->link_components_) {
+            Box box;
+            box.low_ = std::numeric_limits<float>::lowest();
+            box.high_ = std::numeric_limits<float>::max();
+            box.shape_ = {6};
+            box.datatype_ = DataType::Float32;
+            observation_space[link_component.first] = std::move(box);
+        }
+    }
+
+    std::map<std::string, Box> camera_sensor_observation_space = camera_sensor_->getObservationSpace(observation_components);
+    for (auto& camera_sensor_observation_space_component : camera_sensor_observation_space) {
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getStepInfoSpace() const
+{
+    return {};
+}
+
+void UrdfBotAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "joint")) {
```

Comment:
our existing `ACTION_COMPONENTS` have verb-noun pairs in their names like `set_pose`, `apply_voltage`, etc, so let's call this action component `control_joints` for consistency

Reply:
fixed